File: api/application/src/main/java/net/corda/v5/application/flows/CordaInject.java
package net.corda.v5.application.flows;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * The {@link CordaInject} annotation should be used within {@link Flow}s to inject Corda's platform services.
 * <p>
 * The annotation is placed onto a {@link Flow}'s fields/properties and not on constructor parameters.
 * <p>
 * Example usage:
 * <ul>
 * <li>Kotlin:<pre>{@code
 * class MyFlow : ClientStartableFlow {
 *
 *     @CordaInject
 *     lateinit var flowEngine: FlowEngine
 *
 *     @CordaInject
 *     lateinit var flowMessaging: FlowMessaging
 *
 *     @Suspendable
 *     override fun call(requestBody: RestRequestBody): String {
 *         ...
 *     }
 * }
 * }</pre></li>
 * <li>Java:<pre>{@code
 * class MyFlow implements ClientStartableFlow {
 *
 *     @CordaInject
 *     public FlowEngine flowEngine;
 *
 *     @CordaInject
 *     public FlowMessaging flowMessaging;
 *
 *     @Suspendable
 *     @Override
 *     public String call(RestRequestBody requestBody) {
 *         ...
 *     }
 * }
 * }</pre></li></ul>
 *
 * The properties annotated with {@link CordaInject} can be public or private as the reflection that sets them can bypass
 * their visibility modifiers. Making a {@link Flow}'s properties public, internal or package private is convenient for unit
 * testing and does not affect the {@link Flow}'s runtime behaviour.
 * <p>
 * The injected properties cannot be accessed within a {@link Flow}'s constructor and should only be used from within a
 * {@link Flow}'s call method or subsequently executed code.
 */
@Retention(RUNTIME)
@Target(FIELD)
@Documented
public @interface CordaInject {
}

File: application/src/main/kotlin/net/corda/v5/application/flows/CordaInject.kt
package net.corda.v5.application.flows

import kotlin.annotation.AnnotationTarget.FIELD

/**
 * The [CordaInject] annotation should be used within [Flow]s to inject Corda's platform services.
 *
 * The annotation is placed onto a [Flow]'s fields/properties and not on constructor parameters.
 *
 * Example usage:
 *
 * - Kotlin:
 *
 * ```kotlin
 * class MyFlow : RestStartableFlow {
 *
 *     @CordaInject
 *     lateinit var flowEngine: FlowEngine
 *
 *     @CordaInject
 *     lateinit var flowMessaging: FlowMessaging
 *
 *     @Suspendable
 *     override fun call(requestBody: RestRequestBody): String {
 *         ...
 *     }
 * }
 * ```
 *
 * - Java:
 *
 * ```java
 * class MyFlow implements RestStartableFlow {
 *
 *     @CordaInject
 *     public FlowEngine flowEngine;
 *
 *     @CordaInject
 *     public FlowMessaging flowMessaging;
 *
 *     @Suspendable
 *     @Override
 *     public String call(RestRequestBody requestBody) {
 *         ...
 *     }
 * }
 * ```
 *
 * The properties annotated with [CordaInject] can be public or private as the reflection that sets them can bypass
 * their visibility modifiers. Making a [Flow]'s properties public, internal or package private is convenient for unit
 * testing and does not affect the [Flow]'s runtime behaviour.
 *
 * The injected properties cannot be accessed within a [Flow]'s constructor and should only be used from within a
 * [Flow]'s call method or subsequently executed code.
 */
@Target(FIELD)
@MustBeDocumented
annotation class CordaInject